In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings ('ignore')
import matplotlib.pyplot as plt

In [2]:
accident = pd.read_csv('datasets\\accident_data.csv')

In [3]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h3>Extracting date information using pandas date time</h3>

In [4]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors = 'coerce')

In [5]:
accident['Year'] = accident['Accident Date'].dt.year
accident['Month'] = accident['Accident Date'].dt.month
accident['Day'] = accident['Accident Date'].dt.day
accident['DayofWeek'] = accident['Accident Date'].dt.dayofweek #Monday=0 , Sunday=6 

In [65]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 18 columns):
 #   Column                   Non-Null Count   Dtype   
---  ------                   --------------   -----   
 0   Index                    660679 non-null  object  
 1   Accident_Severity        660679 non-null  category
 2   Accident Date            265007 non-null  category
 3   Latitude                 660679 non-null  category
 4   Light_Conditions         660679 non-null  category
 5   District Area            660679 non-null  object  
 6   Longitude                660679 non-null  category
 7   Number_of_Casualties     660679 non-null  int64   
 8   Number_of_Vehicles       660679 non-null  int64   
 9   Road_Surface_Conditions  660679 non-null  category
 10  Road_Type                660679 non-null  category
 11  Urban_or_Rural_Area      660679 non-null  category
 12  Weather_Conditions       660679 non-null  category
 13  Vehicle_Type             660679 non-null  ca

In [64]:
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')
accident['Accident Date'] = accident['Accident Date'].astype('category')
accident['Latitude'] = accident['Latitude'].astype('category')
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')
accident['Longitude'] = accident['Longitude'].astype('category')
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')
accident['Road_Type'] = accident['Road_Type'].astype('category')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')

In [7]:
accident.describe()

Accident Date       Latitude      Longitude  \
count                         265007  660654.000000  660653.000000   
mean   2020-11-23 16:22:04.873683456      52.553866      -1.431210   
min              2019-01-01 00:00:00      49.914430      -7.516225   
25%              2019-12-03 00:00:00      51.490690      -2.332291   
50%              2020-11-07 00:00:00      52.315641      -1.411667   
75%              2021-11-07 00:00:00      53.453452      -0.232869   
max              2022-12-12 00:00:00      60.757544       1.762010   
std                              NaN       1.406922       1.383330   

       Number_of_Casualties  Number_of_Vehicles           Year          Month  \
count         660679.000000       660679.000000  265007.000000  265007.000000   
mean               1.357040            1.831255    2020.407216       6.708374   
min                1.000000            1.000000    2019.000000       1.000000   
25%                1.000000            1.000000    2019.000000       4.000000   
50%                1.000000            2.000000    2020.000000       7.000000   
75%                1.000000            2.000000    2021.000000      10.000000   
max               68.000000           32.000000    2022.000000      12.000000   
std                0.824847            0.715269       1.101126       3.445617   

                 Day      DayofWeek  
count  265007.000000  265007.000000  
mean        6.522028       3.106367  
min         1.000000       0.000000  
25%         4.000000       1.000000  
50%         7.000000       3.000000  
75%        10.000000       5.000000  
max        12.000000       6.000000  
std         3.467063       1.920114

In [8]:
accident.describe().T

count                           mean  \
Accident Date           265007  2020-11-23 16:22:04.873683456   
Latitude              660654.0                      52.553866   
Longitude             660653.0                       -1.43121   
Number_of_Casualties  660679.0                        1.35704   
Number_of_Vehicles    660679.0                       1.831255   
Year                  265007.0                    2020.407216   
Month                 265007.0                       6.708374   
Day                   265007.0                       6.522028   
DayofWeek             265007.0                       3.106367   

                                      min                  25%  \
Accident Date         2019-01-01 00:00:00  2019-12-03 00:00:00   
Latitude                         49.91443             51.49069   
Longitude                       -7.516225            -2.332291   
Number_of_Casualties                  1.0                  1.0   
Number_of_Vehicles                    1.0                  1.0   
Year                               2019.0               2019.0   
Month                                 1.0                  4.0   
Day                                   1.0                  4.0   
DayofWeek                             0.0                  1.0   

                                      50%                  75%  \
Accident Date         2020-11-07 00:00:00  2021-11-07 00:00:00   
Latitude                        52.315641            53.453452   
Longitude                       -1.411667            -0.232869   
Number_of_Casualties                  1.0                  1.0   
Number_of_Vehicles                    2.0                  2.0   
Year                               2020.0               2021.0   
Month                                 7.0                 10.0   
Day                                   7.0                 10.0   
DayofWeek                             3.0                  5.0   

                                      max       std  
Accident Date         2022-12-12 00:00:00       NaN  
Latitude                        60.757544  1.406922  
Longitude                         1.76201   1.38333  
Number_of_Casualties                 68.0  0.824847  
Number_of_Vehicles                   32.0  0.715269  
Year                               2022.0  1.101126  
Month                                12.0  3.445617  
Day                                  12.0  3.467063  
DayofWeek                             6.0  1.920114

In [63]:
accident.isnull().sum()

Index                           0
Accident_Severity               0
Accident Date              395672
Latitude                        0
Light_Conditions                0
District Area                   0
Longitude                       0
Number_of_Casualties            0
Number_of_Vehicles              0
Road_Surface_Conditions         0
Road_Type                       0
Urban_or_Rural_Area             0
Weather_Conditions              0
Vehicle_Type                    0
Year                       395672
Month                      395672
Day                        395672
DayofWeek                  395672
dtype: int64

In [62]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mode()[0])
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mode()[0])
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna(accident['Road_Surface_Conditions'].mode()[0])
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])
accident['Road_Type'] = accident['Road_Type'].fillna(accident['Road_Type'].mode()[0])
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna(accident['Weather_Conditions'].mode()[0])


<h1>Questions and Insights</h1>

<h3>1. Which weather condition has the highest average number of vehicles involved in accidents?</h3>

In [23]:
accident.groupby("Weather_Conditions")["Number_of_Vehicles"].mean()

Weather_Conditions
Fine + high winds        1.779401
Fine no high winds       1.835907
Fog or mist              1.807823
Other                    1.777668
Raining + high winds     1.777951
Raining no high winds    1.831610
Snowing + high winds     1.737853
Snowing no high winds    1.754889
Name: Number_of_Vehicles, dtype: float64

<h3>Insights:According on the given results it shows that the weather condition with the highest average number of vehicles that is involved in accidents is Fine no high winds with an average of 1.835907.
</h3>

<h3>2. What kind of light conditions has the most serious level of accident severity recorded?</h3>

In [14]:
accident.groupby(["Light_Conditions", "Accident_Severity"]).size().unstack()

Accident_Severity            Fatal  Serious  Slight
Light_Conditions                                   
Darkness - lighting unknown     68      794    5622
Darkness - lights lit         1860    19130  108345
Darkness - lights unlit         45      360    2138
Darkness - no lighting        1612     7174   28651
Daylight                      5076    60759  419045

<h3>Insights: According to the results it shows that Daylight had the most serious level of accident severity with a record of 60,759 which is about 10.87% in percentage .
</h3>

<h3>3. Which type of vehicle has the highest number of casualties?</h3>

In [19]:
accident.groupby("Vehicle_Type")["Number_of_Casualties"].size()

Vehicle_Type
Agricultural vehicle                       1947
Bus or coach (17 or more pass seats)      25878
Car                                      497992
Data missing or out of range                  6
Goods 7.5 tonnes mgw and over             17307
Goods over 3.5t. and under 7.5t            6096
Minibus (8 - 16 passenger seats)           1976
Motorcycle 125cc and under                15269
Motorcycle 50cc and under                  7603
Motorcycle over 125cc and up to 500cc      7656
Motorcycle over 500cc                     25657
Other vehicle                              5637
Pedal cycle                                 197
Ridden horse                                  4
Taxi/Private hire car                     13294
Van / Goods 3.5 tonnes mgw or under       34160
Name: Number_of_Casualties, dtype: int64

<h3>Insights: According to the given results we can see that the type of vehicle with the highest number of casualties is Car with a record of 497,992.</h3>

<h3>4. Is there a correlation between the number of vehicles and the number of casualties?</h3>

In [20]:
accident["Number_of_Vehicles"].corr(accident["Number_of_Casualties"])

np.float64(0.22888886126927557)

<h3>Insights: Based on the result it clearly shows that there are no correlation between the no. of vehicles and the no. of casualties.
</h3>

<h3>5. Is there a relation between the Latitude and the number of casualties that is recorded in the dataset?</h3>

In [22]:
accident["Latitude"].corr(accident["Number_of_Casualties"])

np.float64(0.032200686625906395)

<h3>Insights: According to the result of the analysis there is no correlation between the Latitude and the number of casualties in the dataset.</h3>

<h3>6.  What is the average number of vehicles involved in accidents across different accident severity levels?</h3>

In [24]:
accident.groupby("Accident_Severity")["Number_of_Vehicles"].mean()

Accident_Severity
Fatal      1.786976
Serious    1.678327
Slight     1.855864
Name: Number_of_Vehicles, dtype: float64

<h3>Insights: Based on the given results, we can clearly see that Slight level of accident severity had the highest average of vehicles involved and second to it is Serious level severity, while Fatal level of severity had the lowest average.
</h3>

<h3>7.What road type is most involved accident?</h3>

In [25]:
accident['Road_Type'].value_counts()

Road_Type
Single carriageway    496663
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

<h3>Insights: Based on the given results,single carriageways are the most involved road type in accidents.</h3>

<h3>8.Which district has the highest average casualties per accident?</h3>

In [26]:
accident.groupby('District Area')['Number_of_Casualties'].mean().sort_values(ascending=False).head(10)

District Area
Blaeu Gwent          1.693694
West Somerset        1.613990
Derwentside          1.606186
Chester-le-Street    1.603306
Durham               1.580431
Blaenau Gwent        1.579310
Fylde                1.569350
Hambleton            1.568523
Merthyr Tydfil       1.568182
West Lancashire      1.566341
Name: Number_of_Casualties, dtype: float64

<h3>Insights:According to the records that have the highest average casualties per accident of District area is Blaeu Gwent at 1.69, indicating that casualties in this district usually affect more people than in other districts.</h3>

<h3>9. Which vehicle type is most involved in accidents?</h3>

In [27]:
accident[accident['Accident_Severity'] == 'Fatal']['Vehicle_Type'].value_counts().head(5)

Vehicle_Type
Car                                     6577
Van / Goods 3.5 tonnes mgw or under      467
Motorcycle over 500cc                    339
Bus or coach (17 or more pass seats)     325
Goods 7.5 tonnes mgw and over            216
Name: count, dtype: int64

<h3>Insights:Based on the result,Car was the most frequently involved in accidents.</h3>

<h3>10. Which road type in rural areas had the highest number of recorded accidents?</h3>

In [28]:
accident.groupby(['Road_Type','Urban_or_Rural_Area']).size().unstack()

Urban_or_Rural_Area     Rural  Unallocated     Urban
Road_Type                                           
Dual carriageway      48715.0          1.0   50708.0
One way street         1193.0          NaN   12366.0
Roundabout            15545.0          1.0   28446.0
Single carriageway   169243.0          9.0  327411.0
Slip road              4294.0          NaN    2747.0

<h3>Based on the result of the analysis we can see that the Single Carriageway had the highest number of recorded accidents in the rural area.
</h3>

<h3>11. Which weather condition has the lowest fatal level of accident severity?</h3>

In [30]:
accident.groupby(["Weather_Conditions", "Accident_Severity"]).size().unstack()

Accident_Severity      Fatal  Serious  Slight
Weather_Conditions                           
Fine + high winds        175     1245    7134
Fine no high winds      7207    73285  454521
Fog or mist               82      483    2963
Other                    165     1801   15184
Raining + high winds     145     1261    8209
Raining no high winds    848     9468   69380
Snowing + high winds       3      109     773
Snowing no high winds     36      565    5637

<h3> Based on the given results we can find that the weather condition with the lowest fatal accident severity is Snowing + high winds with only resulted a 3 fatal accident.</h3>

<h3>12.Which date has the most recorded of accident?</h3>

In [31]:
accident['Accident Date'].value_counts()

Accident Date
2021-02-11    685
2021-10-06    680
2019-12-06    678
2019-02-01    676
2021-12-04    667
             ... 
2022-01-09    236
2022-12-02    236
2022-04-05    231
2022-02-07    213
2022-01-10    123
Name: count, Length: 576, dtype: int64

<h3>Insights: Based on the result of which date has the most recorded of accident,2021-02-11 has the highest accident with 685 accidents.</h3>

<h3>13. Which road surface condition is linked to the highest number of accidents?</h3>

In [32]:
accident['Road_Surface_Conditions'].value_counts().idxmax()

'Dry'

<h3>Insights:According to the result the Dry got the highest number of accident on road surface condition</h3>

<h3>14. Which district has the most accident ?</h3>

In [33]:
accident['District Area'].value_counts().sort_values(ascending=False)

District Area
Birmingham            13491
Leeds                  8898
Manchester             6720
Bradford               6212
Sheffield              5710
                      ...  
Berwick-upon-Tweed      153
Teesdale                142
Shetland Islands        133
Orkney Islands          117
Clackmannanshire         91
Name: count, Length: 422, dtype: int64

<h3>Insights:Based on the result above the district with the most accident was Birmingham.</h3>

<h3>15. Which among the listed areas has the highest rate accident?</h3>

In [34]:
accident.groupby('District Area')['Number_of_Vehicles'].mean().sort_values(ascending=False)

District Area
Bolsover            2.040501
Runnymede           2.038076
St. Albans          2.032631
Dartford            2.031195
South Bucks         2.022239
                      ...   
Aberdeenshire       1.553368
Clackmannshire      1.487500
Western Isles       1.455056
Orkney Islands      1.393162
Shetland Islands    1.353383
Name: Number_of_Vehicles, Length: 422, dtype: float64

<h3>Insights: Based on the results, Bolsover has the highest rate accident.</h3>

<h3>16. What has the highest casualties based in Road Surface Conditions?</h3>

In [39]:
accident.groupby('Road_Surface_Conditions')['Number_of_Casualties'].mean().sort_values(ascending=False)

Road_Surface_Conditions
Flood over 3cm. deep    1.495575
Wet or damp             1.411064
Snow                    1.359593
Frost or ice            1.348815
Dry                     1.334545
Name: Number_of_Casualties, dtype: float64

<h3>Insights:Based on the result ,which is Flood over 3cm. deep has the highest casualties</h3>

<h3>17. Which year shows the largest year-over-year decrease in accidents?</h3>

In [41]:
accident['Year'].value_counts().sort_index()

Year
2019.0    71867
2020.0    70163
2021.0    66172
2022.0    56805
Name: count, dtype: int64

<h3>Insights:Based on the result the largest year-over-year decrease in accidents is 2019</h3>

<h3>18.Compare average casualties in different light conditions</h3>

In [42]:
accident.groupby("Light_Conditions")["Number_of_Casualties"].mean()

Light_Conditions
Darkness - lighting unknown    1.306447
Darkness - lights lit          1.369374
Darkness - lights unlit        1.331105
Darkness - no lighting         1.541657
Daylight                       1.340309
Name: Number_of_Casualties, dtype: float64

<h3>Insights: Accidents in darker conditions often result in more casualties compared to daylight accident</h3>

<h3>19.Which weather condition has the highest average casualties?</h3>

In [43]:
accident.groupby("Weather_Conditions")["Number_of_Casualties"].mean().sort_values(ascending=False)

Weather_Conditions
Fog or mist              1.452948
Snowing + high winds     1.418079
Raining + high winds     1.416641
Raining no high winds    1.408214
Fine + high winds        1.386018
Other                    1.354869
Fine no high winds       1.347397
Snowing no high winds    1.341776
Name: Number_of_Casualties, dtype: float64

<h3>Insights:Severe weather (fog, snow, heavy rain) usually leads to higher average casualties per accident.</h3>

<h3>20.Which accidents involved both motorcycles and high-speed roads (dual carriageways)?</h3>

In [50]:
accident["Vehicle_Type"].value_counts().sort_values(ascending=False).head(4).to_frame(name="Count")


Count
Vehicle_Type                                
Car                                   497992
Van / Goods 3.5 tonnes mgw or under    34160
Bus or coach (17 or more pass seats)   25878
Motorcycle over 500cc                  25657

<h3>Insights:The result shows that motorcycles on dual carriageways are involved in accidents, with Motorcycle over 500cc having the highest count, indicating that high-speed roads are riskier for powerful motorcycles.</h3>

<h3>21.How are accidents distributed between urban and rural areas?</h3>

In [55]:
a_c = accident["Urban_or_Rural_Area"].value_counts().to_frame(name="Count")
a_c["Percentage"] = (a_c["Count"] / a_c["Count"].sum() * 100).round(1).astype(str) + "%"

a_c

Count Percentage
Urban_or_Rural_Area                   
Urban                421678      63.8%
Rural                238990      36.2%
Unallocated              11       0.0%

<h3>Insights:The results indicates that urban environments account for about 63.8% of all accidents, likely due to higher traffic density and more complex road networks. In contrast, rural areas account for 36.2%, and only a negligible share (0.0%) is unallocated.</h3>

<h3>22. Which road types have the most accidents at night?</h3>

In [56]:
d_r_c = accident[accident["Light_Conditions"].str.contains("Darkness")]["Road_Type"].value_counts().to_frame(name="Count")
total = d_r_c["Count"].sum()
d_r_c["Percentage"] = (d_r_c["Count"] / total * 100).round(1).astype(str) + "%"

d_r_c

Count Percentage
Road_Type                            
Single carriageway  129929      73.9%
Dual carriageway     29072      16.5%
Roundabout           11411       6.5%
One way street        3595       2.0%
Slip road             1792       1.0%

<H3>Insights: According to the result it shows that single carriageways have the most accidents at night (73.7%), followed by dual carriageways (16.6%) and roundabouts (6.5%), indicating that these roads are more prone to nighttime accidents.</H3>

<h3>23. Which road surface conditions caused more cars to have accidents?</h3>

In [58]:
accident[accident['Vehicle_Type']=="Car"]['Road_Surface_Conditions'].value_counts()

Road_Surface_Conditions
Dry                     337860
Wet or damp             140764
Frost or ice             14108
Snow                      4483
Flood over 3cm. deep       777
Name: count, dtype: int64

<h3>Insights:This shows that the most accidents happened on dry roads, followed by wet roads</h3>

<h3>24.  Do multi-vehicle accidents lead to more casualties on average?</h3>

In [60]:
accident.groupby(accident["Number_of_Vehicles"] > 2) ["Number_of_Casualties"].mean()

Number_of_Vehicles
False    1.305799
True     1.804410
Name: Number_of_Casualties, dtype: float64

<h3>Insights: Based on the result Yes,accidents with more than 2 vehicles caused more injuries on average.</h3>

<h3>25How many distributions of accidents by day of the week ?</h3>

In [61]:
accident["Accident Date"].dt.day_name().value_counts()

Accident Date
Saturday     43164
Wednesday    40037
Friday       39822
Thursday     39641
Tuesday      38714
Sunday       35065
Monday       28564
Name: count, dtype: int64

<h3>Insights:The result shows that in the Weekdays, especially Friday, often have the most accidenets</h3>